clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java

```diff
@@ -45,60 +45,82 @@
+                    partitionData.set(PARTITION_KEY_NAME, partitionId);
+                    partitionArray.add(partitionData);
+                }
+                topicData.set(PARTITIONS_KEY_NAME, partitionArray.toArray());
+                topicArray.add(topicData);
             }
-            topicData.set(PARTITIONS_KEY_NAME, partitionArray.toArray());
-            topicArray.add(topicData);
-        }
-        struct.set(TOPICS_KEY_NAME, topicArray.toArray());
+            struct.set(TOPICS_KEY_NAME, topicArray.toArray());
+        } else
+            struct.set(TOPICS_KEY_NAME, null);
+
         this.groupId = groupId;
         this.partitions = partitions;
     }
 
     public OffsetFetchRequest(Struct struct) {
         super(struct);
-        partitions = new ArrayList<>();
-        for (Object topicResponseObj : struct.getArray(TOPICS_KEY_NAME)) {
-            Struct topicResponse = (Struct) topicResponseObj;
-            String topic = topicResponse.getString(TOPIC_KEY_NAME);
-            for (Object partitionResponseObj : topicResponse.getArray(PARTITIONS_KEY_NAME)) {
-                Struct partitionResponse = (Struct) partitionResponseObj;
-                int partition = partitionResponse.getInt(PARTITION_KEY_NAME);
-                partitions.add(new TopicPartition(topic, partition));
+
+        Object[] topicArray = struct.getArray(TOPICS_KEY_NAME);
+        if (topicArray != null) {
+            partitions = new ArrayList<>();
+            for (Object topicResponseObj : struct.getArray(TOPICS_KEY_NAME)) {
+                Struct topicResponse = (Struct) topicResponseObj;
+                String topic = topicResponse.getString(TOPIC_KEY_NAME);
+                for (Object partitionResponseObj : topicResponse.getArray(PARTITIONS_KEY_NAME)) {
+                    Struct partitionResponse = (Struct) partitionResponseObj;
+                    int partition = partitionResponse.getInt(PARTITION_KEY_NAME);
+                    partitions.add(new TopicPartition(topic, partition));
+                }
             }
-        }
+        } else
+            partitions = null;
+
         groupId = struct.getString(GROUP_ID_KEY_NAME);
     }
 
     @Override
     public AbstractResponse getErrorResponse(int versionId, Throwable e) {
         Map<TopicPartition, OffsetFetchResponse.PartitionData> responseData = new HashMap<>();
 
-        for (TopicPartition partition: partitions) {
-            responseData.put(partition, new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET,
-                    OffsetFetchResponse.NO_METADATA,
-                    Errors.forException(e).code()));
+        short errorCode = Errors.NONE.code();
+        short errorCodeThrown = Errors.forException(e).code();
+
+        // there is no need to build responseData for version 2 and later if there is a top-level error
+        if (versionId >= 2 && !OffsetFetchResponse.PARTITION_ERRORS.contains(errorCodeThrown))
```

Comment:
So even the older versions will have an error code at the top level? This would change it to
```
if (versionId < 2 && partitions != null) {
    for (TopicPartition partition: partitions) {
        responseData.put(partition, new OffsetFetchResponse.PartitionData(
                OffsetFetchResponse.INVALID_OFFSET,
                OffsetFetchResponse.NO_METADATA,
                errorThrown.code()));
    }
}

switch (versionId) {
    case 0:
    case 1:
    case 2:
        return new OffsetFetchResponse(responseData, errorThrown, versionId);
    default:
        throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
                versionId, this.getClass().getSimpleName(), ProtoUtils.latestVersion(ApiKeys.OFFSET_FETCH.id)));
}
```